### PR TITLE
Feature/reconnect ws after being idle

### PR DESCRIPTION
**Pull request by AndzejIlj**

### CODE REVIEW

File: services/app/src/helpers/openWsConnection.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import wsWrapper from "./wsWrapper";
+
+export const openWsConnection = (url) => {
```
Comment (jvalatkevicius) [inline services/app/src/helpers/openWsConnection.js:3]: Are you planning on reusing this?
If no, move it into the ws-context file as a separate function.

File: services/app/src/contexts/ws-context.jsx [hunk 2]
```diff
@@ -59,6 +59,15 @@ const WebSocketsContextProvider = ({ children }) => {
     }
   }, [ws])
 
+  window.onfocus = () => {
+    if(ws?.readyState === wsReadyStates.CLOSED) {
+      const webSocket = openWsConnection(url);
+      webSocket.addEventListener('open', () => {
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ws-context.jsx:65]: There is code duplication here.

Update openWsConnection to accept `onOpen` callback.

e.g.
```
openWsConnection(url, {
   onOpen: ws => setWs(ws),
})
```

File: services/app/src/contexts/ws-context.jsx [hunk 3]
```diff
@@ -37,8 +38,7 @@ const WebSocketsContextProvider = ({ children }) => {
       return;
     }
 
-    const WebSockets = wsWrapper(WebSocket);
-    const webSocket = new WebSockets(url);
+    const webSocket = openWsConnection(url);
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ws-context.jsx:41]: use new openWsConnection interface as described below.

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/components/team/Team.jsx [hunk 4]
```diff
@@ -18,15 +18,15 @@ const Team = ({ players, skipMove, children, removePlayer, playerCount }) => {
 
       <div className="team-list-items">
         <div className="team-list-item align-c rubik-font" data-testid="list-current-player">{children}</div>
-        {players.map((player, index) => (
+        {players?.map((player, index) => (
```
Comment (jvalatkevicius) [inline services/app/src/components/team/Team.jsx:21]: why there are so many optional chains here?

File: services/app/src/constants/appConstants.js [hunk 5]
```diff
@@ -1 +1 @@
-export const PING_INTERVAL_DURATION = 30 * 1000;
\ No newline at end of file
+export const PING_INTERVAL_DURATION = 5 * 1000;
```
Comment (jvalatkevicius) [inline services/app/src/constants/appConstants.js:1]: Why this changed?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/app/src/contexts/ws-context.jsx [hunk 6]
```diff
@@ -37,15 +47,10 @@ const WebSocketsContextProvider = ({ children }) => {
       return;
     }
 
-    const WebSockets = wsWrapper(WebSocket);
-    const webSocket = new WebSockets(url);
-
-    webSocket.addEventListener('open', () => {
-      setWs(webSocket);
-    });
+    openWsConnection();
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ws-context.jsx:50]: update as epxplained

File: services/app/src/utils/wsDebugMessages.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
 /* eslint-disable no-console */
 export const wsDebugMessages = (ws) => {
+  if(!ws) return;
```
Comment (jvalatkevicius) [inline services/app/src/utils/wsDebugMessages.js:3]: this wont be needed after updating onOpenConnection

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

**Review by jvalatkevicius — DISMISSED** [review]

File: services/api/src/messaging/players.service.ts [hunk 8]
```diff
@@ -89,9 +89,12 @@ export const clearBoard = (ws: GameWebSocket): void => {
   publishAllPlayers(ws.roomId);
 };
 
-export const checkIfUserAlreadyExists = (ws: GameWebSocket): void => {
+export const checkIfUserAlreadyExists = (
+  ws: GameWebSocket,
+  id: string,
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:94]: rename this to `playerId`

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/api/src/messaging/players.service.ts [hunk 9]
```diff
@@ -84,21 +84,26 @@ const setDefaultStatusForPlayers = (ws: GameWebSocket): void => {
 export const clearBoard = (ws: GameWebSocket): void => {
   gameService.clearBoard(ws.roomId);
   setDefaultStatusForPlayers(ws);
+  sendMessage(ws, MessageType.SetMyTurn, null);
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:87]: why is this needed?

File: services/api/src/messaging/players.service.ts [hunk 10]
```diff
@@ -84,21 +84,26 @@ const setDefaultStatusForPlayers = (ws: GameWebSocket): void => {
...
+  const myTurn = gameService.findMoveByPlayerId(ws.roomId, playerId);
 
   if (myTurn) {
     players.set(ws, {
       ...players.get(ws),
       status: PlayerStatus.FigurePlaced,
     });
-    publish(ws.roomId, { type: MessageType.SetMyTurn, payload: myTurn });
+
+    sendMessage(ws, MessageType.SetMyTurn, myTurn);
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:106]: This action does not fit in this method.
I would expect a boolean returned from this method, instead it does not return anything and has side effects.

File: services/api/src/messaging/players.service.ts [hunk 11]
```diff
@@ -209,20 +210,17 @@ const publishBoard = (roomId: string) => {
...
+}
 
 export const subscribe = (ws: GameWebSocket, newPlayer: Player): void => {
   const players = getPlayers(ws.roomId);
   logger.info(`New player "${newPlayer.name}" joined the game.`);
   players.set(ws, newPlayer);
 
-  checkIfUserAlreadyExists(ws);
+  checkIfUserAlreadyExists(ws, newPlayer.id);
+  publishFinalBoard(ws, players);
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:223]: This does not seem like a good place for this method.

File: services/app/src/contexts/ChessBoardContext.jsx [hunk 12]
```diff
@@ -64,7 +64,7 @@ const ChessBoardContextProvider = ({ children }) => {
     } else {
       clearBoardItems();
     }
-  }, [turns, isAllTurnsMade]);
+  }, [turns]);
```
Comment (jvalatkevicius) [inline services/app/src/contexts/ChessBoardContext.jsx:67]: Why one dependency went missing?

File: services/api/src/game/__tests__/game-room.service.test.ts [hunk 13]
```diff
@@ -8,7 +8,13 @@ jest.mock('uuid', () => ({
 jest.mock('../../logger');
 
 describe('gameRoomService', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
   const roomId = 'abcd-1234';
+  const roomIdToDelete = 'abcd-5678';
```
Comment (jvalatkevicius) [inline services/api/src/game/__tests__/game-room.service.test.ts:16]: remove this.

File: services/api/src/game/__tests__/game-room.service.test.ts [hunk 14]
```diff
@@ -20,22 +26,24 @@ describe('gameRoomService', () => {
...
-    },
-  );
+  it('should not clean up room, because there are less than 1000', () => {
+    gameRoomService.getOrCreateRoom(roomId);
+    gameRoomService.cleanUp();
+    expect(gameRoomService.getClients(roomId)).toBeDefined();
+  });
 
-  it('should print rooms', async () => {
+  it('should clean up the room, because there are more than 1000', () => {
```
Comment (jvalatkevicius) [inline services/api/src/game/__tests__/game-room.service.test.ts:41]: update test to be independent on other tests.

Comment (jvalatkevicius) [reply services/api/src/game/__tests__/game-room.service.test.ts:41]:
```
  it('should clean up the room, because there are more than 1000', () => {
    for (let i = 0; i < 1001; i++) {
      gameRoomService.getOrCreateRoom(`${i}`);
    }
    expect(gameRoomService.rooms.size).toBe(1001);
    gameRoomService.cleanUp();
    expect(gameRoomService.rooms.size).toBe(1000);
  });```

File: services/api/src/game/game-room.service.ts [hunk 15]
```diff
@@ -52,10 +52,15 @@ export const printRooms = (): void => {
...
-  if (getClients(id)?.size === 0) {
-    logger.info(`Closing game room ${id}`);
-    rooms.get(id)?.server.close();
-    rooms.delete(id);
+export const cleanUp = (): void => {
+  if(rooms.size >= 1000) {
+    for (const key of rooms.keys()) {
+      if (getClients(key).size === 0) {
+        logger.info(`Closing game room ${key}`);
+        rooms.get(key).server.close();
```
Comment (jvalatkevicius) [inline services/api/src/game/game-room.service.ts:60]: why is this line needed?

File: services/api/src/messaging/players.service.ts [hunk 16]
```diff
@@ -84,21 +84,26 @@ const setDefaultStatusForPlayers = (ws: GameWebSocket): void => {
...
+  playerId: string,
+): void => {
   const players = getPlayers(ws.roomId);
-  const myTurn = gameService.findMoveByPlayerId(ws.roomId, players.get(ws)?.id);
+  const myTurn = gameService.findMoveByPlayerId(ws.roomId, playerId);
 
   if (myTurn) {
     players.set(ws, {
       ...players.get(ws),
       status: PlayerStatus.FigurePlaced,
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:103]: Why is this needed at all?

File: services/api/src/messaging/players.service.ts (hunk 10 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:106]: Why do we need to have info about previous game. I think we just reset everything going into the next game.

File: services/api/src/messaging/players.service.ts (hunk 9 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:87]: Why do we need to have info about previous game. I think we just reset everything going into the next game.

Comment (jvalatkevicius) [reply services/api/src/messaging/players.service.ts:87]: The same question remains. What is the reason behind having this piece of code when board is cleared? If its not necessary, we should remove it.

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

**Review by jvalatkevicius — APPROVED** [review]

**Review by jvalatkevicius — APPROVED** [review]
